src/ucp/core/ucp_am.c

```diff
@@ -772,11 +859,27 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_proto_progress_am_rndv_rts, (self),
                  uct_pending_req_t *self)
 {
     ucp_request_t *sreq = ucs_container_of(self, ucp_request_t, send.uct);
+    ucs_status_t status;
+    ucp_request_t sreq_tmp;
 
+    sreq_tmp.send.msg_proto.am.flags  = sreq->send.msg_proto.am.flags;
+    sreq_tmp.send.msg_proto.am.header = sreq->send.msg_proto.am.header;
     /* RTS consists of: AM RTS header, packed rkeys and user header */
-    return ucp_rndv_send_rts(sreq, ucp_am_rndv_rts_pack,
-                             sizeof(ucp_rndv_rts_hdr_t) +
-                                     sreq->send.msg_proto.am.header_length);
+    status = ucp_rndv_send_rts(sreq, ucp_am_rndv_rts_pack,
+                               sizeof(ucp_rndv_rts_hdr_t) +
+                                       sreq->send.msg_proto.am.header_length);
+    if (status == UCS_OK) {
+        ucp_am_put_user_header(&sreq_tmp);
+        return UCS_OK;
+    } else if (status == UCS_ERR_NO_RESOURCE) {
+        status = ucp_am_copy_user_header_to_internal_buf(sreq);
+        if (ucs_unlikely(status != UCS_OK)) {
+            return status;
```

Comment:
in general we can't return error codes from sending functions, plz check assert [here](https://github.com/openucx/ucx/blob/master/src/ucp/core/ucp_request.inl#L346)
relevant to other places as well.
I think we need to complete request with error if copying header failed and return `UCS_OK`
@evgeny-leksikov, wdyt?

Reply:
see `ucp_proto_request_abort` and its usage, I guest it should be similar here

Reply:
I see what you mean. Although this error (out of memory) is fatal we still need to abort the operation gracefully?

Reply:
If you are sure it's fatal, we have `ucs_fatal` macro for such cases :)
since you are trying to return error, need to handle it correclty.

Reply:
If i'm not mistaken this send functions are called from ucp_request_try_send, and ucp_request_try_send call ucs_fatal if the returned status is not as expected...